src/cunumeric/scan/scan_local.cu

```diff
@@ -0,0 +1,132 @@
+/* Copyright 2021-2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/scan/scan_local.h"
+#include "cunumeric/scan/scan_local_template.inl"
+
+#include <thrust/scan.h>
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include "cunumeric/cuda_help.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <typename RES>
+static __global__ void __launch_bounds__(THREADS_PER_BLOCK, MIN_CTAS_PER_SM)
+  lazy_kernel(RES* out, RES* sum_val)
+{
+  const size_t idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (idx >= 1) return;
+  sum_val[0] = out[0];
+}
+
+template <ScanCode OP_CODE, LegateTypeCode CODE, int DIM>
+struct ScanLocalImplBody<VariantKind::GPU, OP_CODE, CODE, DIM> {
+  using OP  = ScanOp<OP_CODE, CODE>;
+  using VAL = legate_type_of<CODE>;
+
+  void operator()(OP func,
+                  const AccessorWO<VAL, DIM>& out,
+                  const AccessorRO<VAL, DIM>& in,
+                  Array& sum_vals,
+                  const Pitches<DIM - 1>& pitches,
+                  const Rect<DIM>& rect) const
+  {
+    auto outptr = out.ptr(rect.lo);
+    auto inptr  = in.ptr(rect.lo);
+    auto volume = rect.volume();
+
+    auto stride = rect.hi[DIM - 1] - rect.lo[DIM - 1] + 1;
+
+    Point<DIM> extents = rect.hi - rect.lo + Point<DIM>::ONES();
+    extents[DIM - 1]   = 1;  // one element along scan axis
+
+    auto sum_valsptr = sum_vals.create_output_buffer<VAL, DIM>(extents, true);
+
+    for (uint64_t index = 0; index < volume; index += stride) {
+      // RRRR depending on stride and volume this should either call multiple streams
+      // RRRR or use a cub version (currently not implemented)
+      thrust::inclusive_scan(
+        thrust::device, inptr + index, inptr + index + stride, outptr + index, func);
+      // get the corresponding ND index with base zero to use for sum_val
+      auto sum_valp = pitches.unflatten(index, rect.lo) - rect.lo;
+      // only one element on scan axis
+      sum_valp[DIM - 1] = 0;
+      // write out the partition sum
+      lazy_kernel<<<1, THREADS_PER_BLOCK>>>(&outptr[index + stride - 1], &sum_valsptr[sum_valp]);
```

Comment:
I didn't add streams or any sort of optimization to the GPU version since my next step in the next PR are rewriting it to not use thrust. That should hopefully remove the need for streams, as well as the obvious inefficiencies of how sum_vals is updated.

Reply:
again, let's use a custom stream. there are several reasons to avoid the default stream. plus, add some `CHECK_CUDA_STREAM` macro for debugging purposes. (please look at other CUDA tasks for an example)

Reply:
27108663382755fb23aeebcfc9977cf7b5f3e014